### PR TITLE
Frontier precompiles

### PR DESCRIPTION
**Pull request by olegnn**
- Implement `MetaStorageReader` and `RawStorageReader` precompiles to allow reading any pallet storage entry from within `EVM`.
- Several minor improvements.

### CODE REVIEW

File: pallets/core_mods/src/storage_reader_tests.rs [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        );
+
+        let did = Did([2u8; 32]);
+        let details: StoredDidDetails<Test> =
+            WithNonce::new_with_nonce(OnChainDidDetails::new(IncId::from(1u32), 2, 3), 5).into();
+        Dids::insert(did, &details);
+
+        let input = Input::new("DIDModule", "Dids", MapKey::new(did), Params::None);
+        assert_returned_value!(
+            MetaStorageReader::<Test>::execute(&input.encode(), Some(10_000_000), DUMMY_CTX),
```
Comment (lovesh) [inline pallets/core_mods/src/storage_reader_tests.rs:104]: Why does this fail?

Comment (olegnn) [reply pallets/core_mods/src/storage_reader_tests.rs:104]: Attempt to encode invalid entity (`Did` instead of `StoredDidDetails `)

Comment (lovesh) [reply pallets/core_mods/src/storage_reader_tests.rs:108]: Then why does [this](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R139) succeed? Before both statements, you have a DID being inserted.

Comment (olegnn) [reply pallets/core_mods/src/storage_reader_tests.rs:108]: [Here](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R109) there's an attempt to decode details as `Did`, that's why it fails.
In a [successful test](https://github.com/docknetwork/dock-substrate/pull/118/files#diff-16572ab34bcf31e1999df400d79ba58cd6496d2fddfcfbf0c0f0f8ba6d07f763R142) it decodes details as details, that's why it succeeds.
> Before both statements, you have a DID being inserted.

In both tests firstly we insert `DidDetails` in a map `Did` -> `DidDetails`. However, in the failed test we attempt to decode inserted did details as `Did` while in the successful test we decode it correctly.

Comment (olegnn) [reply pallets/core_mods/src/storage_reader_tests.rs:108]: Yes, that's the reason.

Comment (lovesh) [reply pallets/core_mods/src/storage_reader_tests.rs:108]: Ok

**Review by lovesh — APPROVED** [review]

**Pull request merged**